src/core/server/models/story/story.ts

```diff
@@ -727,3 +742,306 @@ export async function retrieveStorySections(
+        closedAt: now,
+        updatedAt: now,
+      },
+    },
+    {
+      returnOriginal: false,
+    }
+  );
+
+  return result.value;
+}
+
+const getCommentCounts = (options: {
+  story: Readonly<Story>;
+  negate: boolean;
+}) => {
+  const {
+    story: { commentCounts },
+    negate,
+  } = options;
+  const multiplier = negate ? -1 : 1;
+
+  const result: RelatedCommentCounts = createEmptyRelatedCommentCounts();
+
+  if (commentCounts.action) {
+    for (const key in commentCounts.action) {
+      if (result.action.hasOwnProperty.call(commentCounts.action, key)) {
+        const value = commentCounts.action[key];
+        result.action[key] = value * multiplier;
+      }
+    }
+  }
+
+  if (commentCounts.moderationQueue) {
+    result.moderationQueue.total =
+      commentCounts.moderationQueue.total * multiplier;
+
+    let key: keyof typeof commentCounts.moderationQueue.queues;
+    for (key in commentCounts.moderationQueue.queues) {
+      if (
+        result.moderationQueue.queues.hasOwnProperty.call(
+          commentCounts.moderationQueue.queues,
+          key
+        )
+      ) {
+        const value = commentCounts.moderationQueue.queues[key];
+        result.moderationQueue.queues[key] = value * multiplier;
+      }
+    }
+  }
+
+  if (commentCounts.status) {
+    let key: keyof typeof commentCounts.status;
+    for (key in commentCounts.status) {
+      if (result.status.hasOwnProperty.call(commentCounts.status, key)) {
+        const value = commentCounts.status[key];
+        result.status[key] = value * multiplier;
+      }
+    }
+  }
+
+  return result;
+};
+
+export async function archiveStory(
+  mongo: MongoContext,
+  redis: AugmentedRedis,
+  tenantID: string,
+  id: string
+) {
+  const targetStory = await collection(mongo.main).findOne({ id, tenantID });
+  if (!targetStory) {
+    throw new StoryNotFoundError(id);
+  }
+
+  if (targetStory.isArchived) {
+    return;
+  }
+
+  const targetComments = await comments(mongo.main)
+    .find({ tenantID, storyID: id })
+    .toArray();
+  const targetCommentActions = await commentActions(mongo.main)
+    .find({
+      tenantID,
+      storyID: id,
+    })
+    .toArray();
+  const targetCommentIDs = targetComments.map((c) => c.id);
+  const targetCommentModerationActions = await commentModerationActions(
+    mongo.main
+  )
+    .find({
+      tenantID,
+      commentID: { $in: targetCommentIDs },
+    })
+    .toArray();
+
+  if (targetComments && targetComments.length > 0) {
+    await archivedComments(mongo.archive).insertMany(targetComments);
+  }
+  if (targetCommentActions && targetCommentActions.length > 0) {
+    await archivedCommentActions(mongo.archive).insertMany(
+      targetCommentActions
+    );
+  }
+  if (
+    targetCommentModerationActions &&
+    targetCommentModerationActions.length > 0
+  ) {
+    await archivedCommentModerationActions(mongo.archive).insertMany(
+      targetCommentModerationActions
+    );
+  }
```

Comment:
I have a bit of an issue with how _all_ this data is loaded into memory and only then dumped out to each collection. I'd rather instead see something where you could batch these into more manageable chunks. This would be why the "comments" model could instead do something like:

```
cursor = < create cursor of comments to archive >
batch = []
for comment in cursor:
  batch.push(comment)
  if batch.length >= BATCH_MAX:
    archivedComments.insertMany(batch);
    batch = []
if batch.length > 0:
  archivedComments.insertMany(batch);
  batch = []
```

That way the documents are handed over in reasonable batches. `toArray` does this batching internally, but it just keeps going until the cursor is consumed.

Reply:
This has been batched